File: appli-web/src/app/features/nutrition/food-catalogue.ts
/**
 * Logique de liste/filtre du catalogue d'aliments (Catalogue page + FoodPickerSheet).
 *
 * Bouts communs extraits du FoodPickerSheet (NUTRITION_DESIGN §5.3) pour être réutilisés par la
 * page Catalogue dédiée et testés sans Angular : regroupement Favoris / Récents / Tous,
 * recherche à plat, et résolution des aliments récemment consommés depuis les entries.
 */
import { LocalFood } from '@core/models/food.model';
import { LocalMealEntry } from '@core/models/meal-entry.model';
import { effectiveFoodKcal } from './food-kcal';
import { MICRO_SHORT, MICRO_TARGETS, type MicroNutrients } from './micros';
import { MACRO_ABBR, MACRO_KEYS, MACRO_LABEL, MACRO_UNIT, type MacroKey } from './macro-colors';
import { realmOf, type FoodRealm } from './food-category';

/** Bloc de la liste catalogue (titre vide = liste à plat de recherche, sans en-tête). */
export interface FoodGroup {
  title: string;
  foods: LocalFood[];
}

/** Options de regroupement (la page Catalogue peut afficher en plus les aliments archivés). */
export interface FoodGroupOptions {
  /** Inclure les aliments archivés dans un bloc « Archivés » (catalogue) ; sinon ils sont masqués. */
  showArchived?: boolean;
  /** Nombre max d'aliments récents affichés (par défaut 8, comme la sheet d'ajout). */
  recentLimit?: number;
  /** Filtres de seuil macros/micros (combinés en ET), cumulables avec la recherche texte. */
  thresholds?: readonly NutrientThreshold[];
  /** Facette par règne (origine dérivée du groupe) ; null/absent = tous règnes. */
  realm?: FoodRealm | null;
  /** Facette par groupe : codes acceptés (Compléments = 2 codes) ; null/vide = tous groupes. */
  groupCodes?: readonly string[] | null;
}

// ----------------------------- Recherche multi-critères (seuils macros/micros) ----------------

/** Opérateur d'un seuil sur un nutriment (par 100 g) : « au moins » (≥) ou « au plus » (≤). */
export type ThresholdOp = 'gte' | 'lte';

/** Clé d'un nutriment filtrable par seuil (per-100 g) : kcal + 4 macros + les 10 micros. */
export type NutrientKey =
  | 'kcalPer100g'
  | 'proteinPer100g'
  | 'carbsPer100g'
  | 'fatPer100g'
  | 'fiberPer100g'
  | keyof MicroNutrients;

/** Un filtre de seuil actif : nutriment + opérateur + valeur (per 100 g). */
export interface NutrientThreshold {
  key: NutrientKey;
  op: ThresholdOp;
  value: number;
}

/** Métadonnées d'un nutriment proposé dans le panneau de filtres (libellé + unité). */
export interface FilterableNutrient {
  key: NutrientKey;
  label: string;
  /** Libellé abrégé (Prot./Gluc./Lip. ; Ca/Mg/Fe… ; Vit C/D…) pour les affichages denses (filtres compacts). */
  abbr: string;
  unit: string;
}

/** Mapping macro (code couleur) → clé per-100 g du modèle Food. */
const MACRO_FILTER_KEY: Record<MacroKey, NutrientKey> = {
  kcal: 'kcalPer100g',
  protein: 'proteinPer100g',
  carbs: 'carbsPer100g',
  fat: 'fatPer100g',
  fiber: 'fiberPer100g',
};

/** Macros filtrables (kcal + glucides/lipides/protéines/fibres), ordre canonique des macros. */
export const FILTERABLE_MACROS: readonly FilterableNutrient[] = MACRO_KEYS.map((k) => ({
  key: MACRO_FILTER_KEY[k],
  label: MACRO_LABEL[k],
  abbr: MACRO_ABBR[k],
  unit: MACRO_UNIT[k],
}));

/** Micros filtrables (10 vitamines & minéraux), même ordre que les barres du résumé (T3/T4). */
export const FILTERABLE_MICROS: readonly FilterableNutrient[] = MICRO_TARGETS.map((t) => ({
  key: t.key,
  label: t.label,
  abbr: MICRO_SHORT[t.key],
  unit: t.unit,
}));

/**
 * Valeur per-100 g d'un nutriment, pour le filtrage. kcal = kcal *effective* (dérivée des macros
 * selon la source, D12) pour rester cohérent avec la valeur affichée. Micros/fibres absents (null)
 * comptés comme 0 : un seuil « ≥ X » (X > 0) exclut donc naturellement les aliments sans la donnée.
 * Pure, testable sans Angular.
 */
export function foodNutrientValue(food: LocalFood, key: NutrientKey): number {
  if (key === 'kcalPer100g') return effectiveFoodKcal(food);
  return food[key] ?? 0;
}

/** Vrai si l'aliment satisfait TOUS les seuils (ET). Liste vide → vrai. Pure, testable. */
export function passesThresholds(food: LocalFood, thresholds: readonly NutrientThreshold[]): boolean {
  return thresholds.every((t) => {
    const v = foodNutrientValue(food, t.key);
    return t.op === 'gte' ? v >= t.value : v <= t.value;
  });
}

/**
 * Vrai si l'aliment passe la facette catégorie (règne ET groupe, chacune optionnelle, combinées en
 * ET). Le règne est dérivé du groupe (realmOf) ; le filtre par groupe matche sur les codes acceptés.
 * Facettes absentes (null/vide) → toujours vrai. Pure, testable sans Angular.
 */
export function passesCategory(
  food: LocalFood,
  realm: FoodRealm | null | undefined,
  groupCodes: readonly string[] | null | undefined,
): boolean {
  if (realm && realmOf(food.foodGroup) !== realm) return false;
  if (groupCodes && groupCodes.length && !groupCodes.includes((food.foodGroup ?? '').toUpperCase())) {
    return false;
  }
  return true;
}

/**
 * uuids des aliments récemment consommés (entries les plus récentes d'abord, dédupliqués).
 * Pure : trie une copie des entries par `updatedAt` décroissant. Limite par défaut 8.
 */
export function recentFoodUuids(entries: LocalMealEntry[], limit = 8): string[] {
  const seen = new Set<string>();
  const out: string[] = [];
  const sorted = [...entries].sort((a, b) => (b.updatedAt ?? '').localeCompare(a.updatedAt ?? ''));
  for (const e of sorted) {
    if (e.foodUUID && !seen.has(e.foodUUID)) {
      seen.add(e.foodUUID);
      out.push(e.foodUUID);
      if (out.length >= limit) break;
    }
  }
  return out;
}

/**
 * Construit les blocs de la liste catalogue. Recherche texte OU seuils OU facette catégorie actifs →
 * liste à plat filtrée (nom/marque ET seuils macros/micros ET règne/groupe, titre vide) ; sinon
 * Favoris puis Récents puis Tous (aliments non archivés), et optionnellement un bloc « Archivés » en
 * fin de liste. Toutes les facettes sont cumulables (ET). Pure, testable.
 */
export function buildFoodGroups(
  foods: LocalFood[],
  recentUuids: string[],
  query: string,
  options: FoodGroupOptions = {},
): FoodGroup[] {
  const showArchived = options.showArchived ?? false;
  const thresholds = options.thresholds ?? [];
  const realm = options.realm ?? null;
  const groupCodes = options.groupCodes ?? null;
  const hasCategory = !!realm || !!(groupCodes && groupCodes.length);
  const active = foods.filter((f) => !f.archived);
  const q = query.trim().toLowerCase();

  // Recherche texte OU seuil OU facette catégorie → liste à plat filtrée (toutes les facettes en ET).
  if (q || thresholds.length || hasCategory) {
    const pool = showArchived ? foods : active;
    const filtered = pool.filter(
      (f) =>
        (!q || f.name.toLowerCase().includes(q) || (f.brand ?? '').toLowerCase().includes(q)) &&
        passesThresholds(f, thresholds) &&
        passesCategory(f, realm, groupCodes),
    );
    return filtered.length ? [{ title: '', foods: filtered }] : [];
  }

  const favorites = active.filter((f) => f.isFavorite);
  const favSet = new Set(favorites.map((f) => f.uuid));
  const byUuid = new Map(active.map((f) => [f.uuid, f]));
  const recents = recentUuids
    .filter((uuid) => !favSet.has(uuid))
    .map((uuid) => byUuid.get(uuid))
    .filter((f): f is LocalFood => !!f);
  const recentSet = new Set(recents.map((f) => f.uuid));
  const rest = active.filter((f) => !favSet.has(f.uuid) && !recentSet.has(f.uuid));

  const groups: FoodGroup[] = [];
  if (favorites.length) groups.push({ title: 'Favoris', foods: favorites });
  if (recents.length) groups.push({ title: 'Récents', foods: recents });
  if (rest.length) groups.push({ title: 'Tous', foods: rest });
  if (showArchived) {
    const archived = foods.filter((f) => f.archived);
    if (archived.length) groups.push({ title: 'Archivés', foods: archived });
  }
  return groups;
}

File: appli-android/app/src/main/java/com/example/sportapp/feature/nutrition/domain/FoodCatalogue.kt
package com.example.sportapp.feature.nutrition.domain

import com.example.sportapp.core.data.model.Food
import com.example.sportapp.core.data.model.MealEntry

/**
 * Logique pure du Catalogue d'aliments (A3) — recherche plein-texte + filtres
 * multi-critères par seuil (combinables) + regroupement de la liste. Miroir
 * Android de `appli-web/.../nutrition/food-catalogue.ts`. Sans Compose ni
 * Android, testable en isolation.
 *
 * Les seuils portent sur la valeur per-100 g d'un nutriment (kcal *effective*
 * D12 pour rester cohérent avec l'affichage, macros, et les 10 micros). Tous les
 * critères se combinent en ET. Les labels d'affichage vivent dans strings.xml
 * (politique 18) — ici on ne manipule que des clés et des nombres.
 */

/** Opérateur d'un seuil : « au moins » (≥) ou « au plus » (≤). */
enum class ThresholdOp { GTE, LTE }

/**
 * Nutriment filtrable par seuil (per-100 g) : kcal + 4 macros + 10 micros.
 * Ordre = ordre d'affichage dans le panneau de filtres (macros puis micros).
 * `unit` = symbole d'unité affiché à côté du champ de valeur.
 */
enum class NutrientKey(val unit: String) {
    KCAL("kcal"),
    PROTEIN("g"),
    CARBS("g"),
    FAT("g"),
    FIBER("g"),
    IRON("mg"),
    CALCIUM("mg"),
    MAGNESIUM("mg"),
    ZINC("mg"),
    POTASSIUM("mg"),
    SODIUM("mg"),
    VITAMIN_C("mg"),
    VITAMIN_D("µg"),
    VITAMIN_B12("µg"),
    VITAMIN_A("µg"),
}

/** Un filtre de seuil actif : nutriment + opérateur + valeur (per 100 g). */
data class NutrientThreshold(val key: NutrientKey, val op: ThresholdOp, val value: Float)

/**
 * Valeur per-100 g d'un nutriment, pour le filtrage. kcal = kcal *effective*
 * (dérivée selon la source, D12) pour rester cohérent avec la valeur affichée.
 * Micros/fibres absents (null) comptés comme 0 : un seuil « ≥ X » (X > 0) exclut
 * donc naturellement les aliments sans la donnée. Pur.
 */
fun foodNutrientValue(food: Food, key: NutrientKey): Float = when (key) {
    NutrientKey.KCAL -> effectiveFoodKcal(food)
    NutrientKey.PROTEIN -> food.proteinPer100g
    NutrientKey.CARBS -> food.carbsPer100g
    NutrientKey.FAT -> food.fatPer100g
    NutrientKey.FIBER -> food.fiberPer100g ?: 0f
    NutrientKey.IRON -> food.ironPer100g ?: 0f
    NutrientKey.CALCIUM -> food.calciumPer100g ?: 0f
    NutrientKey.MAGNESIUM -> food.magnesiumPer100g ?: 0f
    NutrientKey.ZINC -> food.zincPer100g ?: 0f
    NutrientKey.POTASSIUM -> food.potassiumPer100g ?: 0f
    NutrientKey.SODIUM -> food.sodiumPer100g ?: 0f
    NutrientKey.VITAMIN_C -> food.vitaminCPer100g ?: 0f
    NutrientKey.VITAMIN_D -> food.vitaminDPer100g ?: 0f
    NutrientKey.VITAMIN_B12 -> food.vitaminB12Per100g ?: 0f
    NutrientKey.VITAMIN_A -> food.vitaminAPer100g ?: 0f
}

/** Vrai si l'aliment satisfait TOUS les seuils (ET). Liste vide → vrai. Pur. */
fun passesThresholds(food: Food, thresholds: List<NutrientThreshold>): Boolean =
    thresholds.all { t ->
        val v = foodNutrientValue(food, t.key)
        if (t.op == ThresholdOp.GTE) v >= t.value else v <= t.value
    }

/** Bloc de la liste catalogue (group null = liste à plat de recherche, sans en-tête). */
enum class CatalogueGroup { RECENTS, FAVORITES, ALL, ARCHIVED }

data class FoodGroupBlock(val group: CatalogueGroup?, val foods: List<Food>)

/**
 * uuids des aliments récemment consommés (entries les plus récentes d'abord, dédupliqués,
 * limite 8 par défaut) — miroir `recentFoodUuids` web. Alimente le bloc « Récents ».
 */
fun recentFoodUuids(entries: List<MealEntry>, limit: Int = 8): List<String> {
    val seen = mutableSetOf<String>()
    val out = mutableListOf<String>()
    for (e in entries.sortedByDescending { it.updatedAt ?: "" }) {
        val id = e.foodUUID ?: continue
        if (seen.add(id)) {
            out += id
            if (out.size >= limit) break
        }
    }
    return out
}

/**
 * Construit les blocs de la liste catalogue. Recherche texte OU seuils actifs →
 * liste à plat filtrée (nom/marque ET seuils macros/micros, sans en-tête) ;
 * sinon Favoris puis Tous (aliments non archivés), et optionnellement un bloc
 * « Archivés » en fin de liste. Recherche et seuils se combinent en ET. Suppose
 * `foods` déjà trié par nom (DAO `ORDER BY name ASC`). Pur.
 */
fun buildCatalogue(
    foods: List<Food>,
    query: String,
    thresholds: List<NutrientThreshold>,
    showArchived: Boolean,
    // uuids récemment consommés (cf. [recentFoodUuids]) → bloc « Récents » en tête (miroir web).
    recentUuids: List<String> = emptyList(),
): List<FoodGroupBlock> {
    val active = foods.filter { !it.archived }
    val q = query.trim()

    // Recherche texte OU seuil → liste à plat filtrée (toutes les facettes en ET).
    if (q.isNotEmpty() || thresholds.isNotEmpty()) {
        val pool = if (showArchived) foods else active
        val filtered = pool.filter { f ->
            (q.isEmpty() ||
                f.name.contains(q, ignoreCase = true) ||
                (f.brand?.contains(q, ignoreCase = true) == true)) &&
                passesThresholds(f, thresholds)
        }
        return if (filtered.isEmpty()) emptyList() else listOf(FoodGroupBlock(null, filtered))
    }

    val favorites = active.filter { it.isFavorite }
    val favSet = favorites.map { it.uuid }.toSet()
    // Récents = consommés récemment HORS favoris (miroir buildFoodGroups web) ; « Tous » exclut les deux.
    val byUuid = active.associateBy { it.uuid }
    val recents = recentUuids.filter { it !in favSet }.mapNotNull { byUuid[it] }
    val recentSet = recents.map { it.uuid }.toSet()
    val rest = active.filter { it.uuid !in favSet && it.uuid !in recentSet }

    val blocks = mutableListOf<FoodGroupBlock>()
    if (recents.isNotEmpty()) blocks += FoodGroupBlock(CatalogueGroup.RECENTS, recents)
    if (favorites.isNotEmpty()) blocks += FoodGroupBlock(CatalogueGroup.FAVORITES, favorites)
    if (rest.isNotEmpty()) blocks += FoodGroupBlock(CatalogueGroup.ALL, rest)
    if (showArchived) {
        val archived = foods.filter { it.archived }
        if (archived.isNotEmpty()) blocks += FoodGroupBlock(CatalogueGroup.ARCHIVED, archived)
    }
    return blocks
}
